partiql-types/src/main/kotlin/org/partiql/value/datetime/DateTime.kt

```diff
@@ -0,0 +1,560 @@
+     * [hours] can be negative.
+     */
+    public fun plusHours(hours: Long): Timestamp
+
+    /**
+     * Returns a [Timestamp] value with the specified number of minutes added.
+     * Year, Month, Day, Hour fields may be changed to ensure the correctness of the result.
+     * [minutes] can be negative.
+     */
+    public fun plusMinutes(minutes: Long): Timestamp
+
+    /**
+     * Returns a [Timestamp] value with the specified number of minutes added.
+     * Year, Month, Day, Hour, Minute fields may be changed to ensure the correctness of the result.
+     * [seconds] can be negative.
+     */
+    public fun plusSeconds(seconds: BigDecimal): Timestamp
+
+    /**
+     * Convenient wrapper for entering long value as [seconds]
+     */
+    public fun plusSeconds(seconds: Long): Timestamp
+
+    /**
+     * Returns a [Timestamp] value with [precision] number of digits in second fraction.
+     */
+    public fun toPrecision(precision: Int): Timestamp
+
+    /**
+     * Returns a [Date] value by extracting [year], [month], [day] field
+     */
+    public fun toDate(): Date = DateTimeValue.date(this.year, this.month, this.day)
+
+    /**
+     * Returns a [Time] value by extracting [hour], [minute], [decimalSecond], and [timeZone].
+     */
+    public fun toTime(): Time
+
+    /**
+     * Comparison method for timestamp value
+     * If both value are [TimestampWithTimeZone]:
+     *    The two value are considered equivalent if they refer to the same point in time.
+     * If both value are timestamp without timezone:
+     *      The two value are consider equivalent if all the fields (exclude precision) are equivalent.
+     *      Another way to interpret this is, two timestamp without timezone are equivalent
+     *          if and only if they are equivalent when converted to the same time zone.
+     * If one value is timestamp with time zone and the other is timestamp without time zone:
+     *      Error.
+     *      One may not directly compare a timestamp value with timezone to a timestamp value without time zone.
+     */
+    public override fun compareTo(other: Timestamp): Int =
+        DateTimeComparator.compareTo(this, other)
+}
+
+/**
+ * Superclass for all implementation representing date value
+ */
+public abstract class DateImpl : Date, Comparable<Date> {
+    public final override fun equals(other: Any?): Boolean {
+        if (this === other) return true
+        if (other?.javaClass != javaClass) return false
+        other as DateImpl
+        if (this.year != other.year) return false
+        if (this.month != other.month) return false
+        if (this.day != other.day) return false
+        return true
+    }
+
+    public final override fun hashCode(): Int =
+        year.hashCode() + month.hashCode() + day.hashCode()
+
+    public final override fun toString(): String =
+        "${this.javaClass.simpleName}(year=$year, month=$month, day=$day)"
+}
```

Comment:
Why not make `Date` the abstract class with these default implementations?

Reply:
The original thought was to accommodate the possibility of having to implement different date system.  Maybe I am overcomplicating this stuff and we would never need this...